# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: netwerk/ipc/SocketProcessBridgeChild.cpp
# Commit: d4004105a04a
# Full Hash: d4004105a04ad9a473930acb9c6a05d90cd9f028
# Author: Byron Campen [:bwc] <docfaraday@gmail.com>
# Date: 2019-02-23 04:13:22
# Regressor Bug: 1521879
# File Overlap Count: 1
# Description:
#   Bug 1521879 - Part 1: IPC-based MediaTransport implementation r=mjf
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D17273
# ==============================================================================

diff -r baecf588657c -r d4004105a04a netwerk/ipc/SocketProcessBridgeChild.cpp
--- a/netwerk/ipc/SocketProcessBridgeChild.cpp	Thu Feb 21 23:16:40 2019 +0000
+++ b/netwerk/ipc/SocketProcessBridgeChild.cpp	Thu Feb 21 20:24:54 2019 +0000
@@ -9,6 +9,7 @@
 #include "mozilla/net/NeckoChild.h"
 #include "nsIObserverService.h"
 #include "nsThreadUtils.h"
+#include "mozilla/dom/PMediaTransportChild.h"
 
 namespace mozilla {
 namespace net {
@@ -47,35 +48,49 @@
 }
 
 // static
-void SocketProcessBridgeChild::EnsureSocketProcessBridge(
-    std::function<void()>&& aOnSuccess, std::function<void()>&& aOnFailure) {
-  MOZ_ASSERT(IsNeckoChild() && gNeckoChild);
+RefPtr<SocketProcessBridgeChild::GetPromise>
+SocketProcessBridgeChild::GetSocketProcessBridge() {
   MOZ_ASSERT(NS_IsMainThread());
 
   if (!gNeckoChild) {
-    aOnFailure();
-    return;
+    return GetPromise::CreateAndReject(nsCString("No NeckoChild!"), __func__);
   }
 
   if (sSocketProcessBridgeChild) {
-    aOnSuccess();
-    return;
+    return GetPromise::CreateAndResolve(sSocketProcessBridgeChild, __func__);
   }
 
-  gNeckoChild->SendInitSocketProcessBridge()->Then(
+  return gNeckoChild->SendInitSocketProcessBridge()->Then(
       GetMainThreadSerialEventTarget(), __func__,
-      [onSuccess = std::move(aOnSuccess), onFailure = std::move(aOnFailure)](
-          Endpoint<PSocketProcessBridgeChild>&& aEndpoint) {
-        if (aEndpoint.IsValid()) {
-          if (SocketProcessBridgeChild::Create(std::move(aEndpoint))) {
-            onSuccess();
-            return;
+      [](NeckoChild::InitSocketProcessBridgePromise::ResolveOrRejectValue&&
+             aResult) {
+        if (!sSocketProcessBridgeChild) {
+          if (aResult.IsReject()) {
+            return GetPromise::CreateAndReject(
+                nsCString("SendInitSocketProcessBridge failed"), __func__);
+          }
+
+          if (!aResult.ResolveValue().IsValid()) {
+            return GetPromise::CreateAndReject(
+                nsCString(
+                    "SendInitSocketProcessBridge resolved with an invalid "
+                    "endpoint!"),
+                __func__);
+          }
+
+          if (!SocketProcessBridgeChild::Create(
+                  std::move(aResult.ResolveValue()))) {
+            return GetPromise::CreateAndReject(
+                nsCString("SendInitSocketProcessBridge resolved with a valid "
+                          "endpoint, "
+                          "but SocketProcessBridgeChild::Create failed!"),
+                __func__);
           }
         }
-        onFailure();
-      },
-      [onFailure = std::move(aOnFailure)](
-          const mozilla::ipc::ResponseRejectReason) { onFailure(); });
+
+        return GetPromise::CreateAndResolve(sSocketProcessBridgeChild,
+                                            __func__);
+      });
 }
 
 SocketProcessBridgeChild::SocketProcessBridgeChild(
@@ -133,5 +148,21 @@
   sSocketProcessBridgeChild = nullptr;
 }
 
+dom::PMediaTransportChild*
+SocketProcessBridgeChild::AllocPMediaTransportChild() {
+  // We don't allocate here: MediaTransportHandlerIPC is in charge of that,
+  // so we don't need to know the implementation particulars here.
+  MOZ_ASSERT_UNREACHABLE(
+      "The only thing that ought to be creating a PMediaTransportChild is "
+      "MediaTransportHandlerIPC!");
+  return nullptr;
+}
+
+bool SocketProcessBridgeChild::DeallocPMediaTransportChild(
+    dom::PMediaTransportChild* aActor) {
+  delete aActor;
+  return true;
+}
+
 }  // namespace net
 }  // namespace mozilla